File: 프로그래머스/3/84021. 퍼즐 조각 채우기/퍼즐 조각 채우기.py
import copy

dx = [-1, 0, 1, 0]
dy = [0, -1, 0, 1]

def dfs(board, a, b, position, num):
    ret = [position]

    for i in range(4):
        nx = a + dx[i]
        ny = b + dy[i]
        if 0 <= nx < len(board) and 0<= ny < len(board):
            if board[ny][nx] == num:
                board[ny][nx] = 2
                ret = ret + dfs(board, nx, ny, [position[0] + dx[i], position[1] + dy[i]], num)
    return ret

def rotate(table, i):
    if not i:
        return table
    new_table = []

    for x in range(len(table)):
        new_table.append([])
        for y in range(len(table)-1, -1, -1):
            new_table[x].append(table[y][x])
    return new_table

def solution(game_board, table):
    answer = 0

    blocks = []
    for y in range(len(game_board)):
        for x in range(len(game_board)):
            if game_board[y][x] == 0:
                game_board[y][x] = 2
                blocks.append(dfs(game_board, x, y, [0, 0], 0))

    for i in range(4):
        table = rotate(table, i)
        table_copy = copy.deepcopy(table)
        for y in range(len(table)):
            for x in range(len(table)):
                if table_copy[y][x] == 1:
                    table_copy[y][x] = 2
                    temp = dfs(table_copy, x, y, [0, 0], 1)
                    if temp in blocks:
                        answer += len(temp)
                        blocks.pop(blocks.index(temp))
                        table = copy.deepcopy(table_copy)
                    else:
                        table_copy = copy.deepcopy(table)
 
    return answer

File: 프로그래머스/3/84021. 퍼즐 조각 채우기/퍼즐 조각 채우기.js
const bfs = (board, x, y, target, visited) => {
    const [r, c] = [board.length, board[0].length]
    
    const q = [{x, y, pos: [0, 0]}];
    visited.add(`${x},${y}`)
    
    const path = [];
    const dx = [-1, 0, 1, 0];
    const dy = [0, -1, 0, 1];

    while (q.length !== 0) {
        const {x, y, pos} = q.shift();
        path.push(pos.join(","));

        for (let i = 0; i < 4; i ++) {
            const nx = x + dx[i];
            const ny = y + dy[i];

            if (!(0 <= nx && nx < c && 0 <= ny && ny < r)) {
                continue;
            }

            if (board[ny][nx] != target) continue;
            const mask = `${nx},${ny}`;
            if (visited.has(mask)) continue;
            visited.add(mask);
            q.push({x:nx, y:ny, pos: [pos[0] + dx[i], pos[1] + dy[i]]});
        }
    }
    return path.join('|')
}

const rotate = (board) => {
    const [r, c] = [board.length, board[0].length]
    
    const newBoard = [];
    
    for (let x = 0; x < c; x ++) {
        newBoard.push([]);
        for (let y = r - 1; y >= 0; y --) {
            newBoard[x].push(board[y][x]);
        }
    }
    return newBoard;
}

const removeBlock = (board, x, y, path) => {
    for (const pos of path.split("|")) {
        const [dx, dy] = pos.split(",").map(Number);
        const nx = x + dx;
        const ny = y + dy;
        board[ny][nx] = 0;
    } 
}


function solution(game_board, table) {
    const [r, c] = [game_board.length, game_board[0].length]
    const emptyAreas = {}
    
    const visited = new Set();
    for (let y = 0; y < r; y ++) {
        for (let x = 0; x < c; x ++) {
            if (game_board[y][x] == 1) continue;
            if (visited.has(`${x},${y}`)) continue;
            const path = bfs(game_board, x, y, 0, visited);
            if (!(path in emptyAreas)) {
                emptyAreas[path] = 0;
            }
            emptyAreas[path] += 1;
        }
    }
    let answer = 0;
    let copyTable = table;
    for (let i = 0; i < 4; i ++) {
        if (i !== 0) {
            copyTable = rotate(copyTable);
        }
        const visited = new Set();
        for (let y = 0; y < r; y ++) {
            for (let x = 0; x < c; x ++) {
                if (copyTable[y][x] === 0) continue; 
                if (visited.has(`${x},${y}`)) continue;
                
                const path = bfs(copyTable, x, y, 1, visited);
                
                if (path in emptyAreas) {
                    answer += path.split("|").length;
                    emptyAreas[path] -= 1;
                    if (emptyAreas[path] === 0) {
                        delete emptyAreas[path];
                    }
                    removeBlock(copyTable, x, y, path);
                }
            }
        }
        
    }
    
    return answer;
}